nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-server/src/main/java/org/apache/nifi/distributed/cache/server/codec/CacheRequestDecoder.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Decode Byte Buffer reading header on initial connection followed by protocol version and cache operations
+     *
+     * @param channelHandlerContext Channel Handler Context
+     * @param byteBuf Byte Buffer
+     * @param objects Decoded Objects
+     */
+    @Override
+    protected void decode(final ChannelHandlerContext channelHandlerContext, final ByteBuf byteBuf, final List<Object> objects) {
+        if (!headerReceived.get()) {
+            byteBuf.readBytes(HEADER_LENGTH);
+            headerReceived.getAndSet(true);
+            log.debug("Header Received [{}]", channelHandlerContext.channel().remoteAddress());
+        }
+
+        if (protocolVersion.get() == 0) {
+            final int clientVersion = byteBuf.readInt();
+            log.debug("Protocol Version [{}] Received [{}]", clientVersion, channelHandlerContext.channel().remoteAddress());
+            final CacheVersionRequest cacheVersionRequest = new CacheVersionRequest(clientVersion);
+            objects.add(cacheVersionRequest);
+        } else {
+            final CacheOperation cacheOperation = readOperation(byteBuf);
+            final Object cacheRequest = readRequest(cacheOperation, byteBuf);
+            if (cacheRequest == null) {
+                log.debug("Cache Operation [{}] request not processed", cacheOperation);
+            } else {
+                objects.add(cacheRequest);
+                // Reset Cache Operation after successful decoding
+                currentOperation.set(null);
+            }
+        }
+    }
+
+    @Override
+    public void exceptionCaught(final ChannelHandlerContext context, final Throwable cause) {
+        log.warn("Request Decoding Failed: Closing Connection [{}]", context.channel().remoteAddress(), cause);
+        context.close();
+    }
+
+    /**
+     * Set Protocol Version based on version negotiated in other handlers
+     *
+     * @param protocolVersion Protocol Version
+     */
+    public void setProtocolVersion(final int protocolVersion) {
+        this.protocolVersion.getAndSet(protocolVersion);
+    }
+
+    /**
+     * Read Request Object based on Cache Operation
+     *
+     * @param cacheOperation Cache Operation
+     * @param byteBuf Byte Buffer
+     * @return Request Object or null when buffer does not contain sufficient bytes
+     */
+    protected Object readRequest(final CacheOperation cacheOperation, final ByteBuf byteBuf) {
+        final int readableBytes = byteBuf.readableBytes();
+        final byte[] body = readableBytes == 0 ? new byte[0] : readBytes(byteBuf);
+        return body == null ? null : new CacheRequest(cacheOperation, body);
+    }
+
+    /**
+     * Read Bytes from buffer based on length indicated
+     *
+     * @param byteBuf Byte Buffer
+     * @return Bytes read or null when buffer does not contain sufficient bytes
+     */
+    protected byte[] readBytes(final ByteBuf byteBuf) {
+        final int length = readBytesLength(byteBuf);
```

Comment:
Is there some sort of buffering implemented, in case the entire request payload is not received in a single netty ByteBuf?  We should be able to handle some small amount of latency, in the case of a heavily loaded network.